File: sdk/javascript/src/nem/Network.js
import ByteArray from '../ByteArray.js';
import { Network as BasicNetwork } from '../Network.js';
import { NetworkTimestamp as BasicNetworkTimestamp, NetworkTimestampDatetimeConverter } from '../NetworkTimestamp.js';
import base32 from '../utils/base32.js';
import { keccak_256 } from '@noble/hashes/sha3';

/**
 * Represents a NEM network timestamp with second resolution.
 */
export class NetworkTimestamp extends BasicNetworkTimestamp {
	/**
	 * Adds a specified number of seconds to this timestamp.
	 * @override
	 * @param {number} count Number of seconds to add.
	 * @returns {NetworkTimestamp} New timestamp that is the specified number of seconds past this timestamp.
	 */
	addSeconds(count) {
		return new NetworkTimestamp(this.timestamp + BigInt(count));
	}
}

/**
 * Represents a NEM address.
 */
export class Address extends ByteArray {
	/**
	 * Byte size of raw address.
	 * @type number
	 */
	static SIZE = 25;

	/**
	 * Length of encoded address string.
	 * @type number
	 */
	static ENCODED_SIZE = 40;

	/**
	 * Creates a NEM address.
	 * @param {Uint8Array|string|Address} addressInput Input string, byte array or address.
	 */
	constructor(addressInput) {
		const extractAddressBytes = () => {
			if ('string' === typeof addressInput)
				return base32.decode(addressInput);

			if (addressInput instanceof Address)
				return addressInput.bytes;

			return addressInput;
		};

		super(Address.SIZE, extractAddressBytes());
	}

	/**
	 * Returns string representation of this object.
	 * @returns {string} String representation of this object.
	 */
	toString() {
		return base32.encode(this.bytes);
	}
}

/**
 * Represents a NEM network.
 */
export class Network extends BasicNetwork {
	/**
	 * NEM main network.
	 * @type Network
	 */
	static MAINNET;

	/**
	 * NEM test network.
	 * @type Network
	 */
	static TESTNET;

	/**
	 * NEM well known networks.
	 * @type Array<Network>
	 */
	static NETWORKS;

	/**
	 * Creates a new network with the specified name, identifier byte and generation hash seed.
	 * @param {string} name Network name.
	 * @param {number} identifier Network identifier byte.
	 * @param {Date} epochTime Network epoch time.
	 */
	constructor(name, identifier, epochTime) {
		super(
			name,
			identifier,
			new NetworkTimestampDatetimeConverter(epochTime, 'seconds'),
			() => keccak_256.create(),
			(addressWithoutChecksum, checksum) => new Address(new Uint8Array([...addressWithoutChecksum, ...checksum])),
			Address,
			NetworkTimestamp
		);
	}
}

Network.MAINNET = new Network('mainnet', 0x68, new Date(Date.UTC(2015, 2, 29, 0, 6, 25)));
Network.TESTNET = new Network('testnet', 0x98, new Date(Date.UTC(2015, 2, 29, 0, 6, 25)));
Network.NETWORKS = [Network.MAINNET, Network.TESTNET];

File: sdk/cpp/symbolchain/nem/Network.py
import base64
import datetime

import sha3

from ..ByteArray import ByteArray
from ..Network import Network as BasicNetwork
from ..NetworkTimestamp import NetworkTimestamp as BasicNetworkTimestamp
from ..NetworkTimestamp import NetworkTimestampDatetimeConverter


class NetworkTimestamp(BasicNetworkTimestamp):
	"""Represents a nem network timestamp with second resolution."""

	def add_seconds(self, count):
		return NetworkTimestamp(self.timestamp + count)


class Address(ByteArray):
	"""Represents a nem address."""

	SIZE = 25
	ENCODED_SIZE = 40

	def __init__(self, address):
		"""Creates an address from a decoded or encoded address."""
		raw_bytes = address
		if isinstance(address, str):
			raw_bytes = base64.b32decode(address)
		elif isinstance(address, Address):
			raw_bytes = address.bytes

		super().__init__(self.SIZE, raw_bytes, Address)

	def __str__(self):
		return base64.b32encode(self.bytes).decode('utf8')

	def __repr__(self):
		return f'Address(\'{str(self)}\')'


class Network(BasicNetwork):
	"""Represents a nem network."""

	def __init__(self, name, identifier, epoch_time):
		"""Creates a new network with the specified properties."""
		super().__init__(name, identifier, NetworkTimestampDatetimeConverter(epoch_time, 'seconds'), Address, NetworkTimestamp)

	def address_hasher(self):
		return sha3.keccak_256()

	def create_address(self, address_without_checksum, checksum):
		return Address(address_without_checksum + checksum)


Network.MAINNET = Network('mainnet', 0x68, datetime.datetime(2015, 3, 29, 0, 6, 25, tzinfo=datetime.timezone.utc))
Network.TESTNET = Network('testnet', 0x98, datetime.datetime(2015, 3, 29, 0, 6, 25, tzinfo=datetime.timezone.utc))
Network.NETWORKS = [Network.MAINNET, Network.TESTNET]
